src/renderer/components/node/CollapsedHandles.tsx

```diff
@@ -0,0 +1,182 @@
+import { NeverType } from '@chainner/navi';
+import { Box } from '@chakra-ui/react';
+import { memo, useMemo } from 'react';
+import { Handle, Node, Position, useReactFlow } from 'reactflow';
+import { useContext, useContextSelector } from 'use-context-selector';
+import { InputId, NodeData } from '../../../common/common-types';
+import {
+    parseSourceHandle,
+    parseTargetHandle,
+    stringifySourceHandle,
+    stringifyTargetHandle,
+} from '../../../common/util';
+import { BackendContext } from '../../contexts/BackendContext';
+import { GlobalVolatileContext } from '../../contexts/GlobalNodeState';
+import { defaultColor, getTypeAccentColors } from '../../helpers/accentColors';
+import { NodeState } from '../../helpers/nodeState';
+import { getBackground } from '../Handle';
+
+interface InputHandleProps {
+    nodeId: string;
+    inputId: InputId;
+    isIterated: boolean;
+    handleId: string;
+}
+
+const InputHandle = memo(({ nodeId, isIterated, inputId, handleId }: InputHandleProps) => {
+    const { edgeChanges, typeState } = useContext(GlobalVolatileContext);
+    const { getEdges, getNode } = useReactFlow();
+
+    const connectedEdge = useMemo(() => {
+        return getEdges().find(
+            (e) => e.target === nodeId && parseTargetHandle(e.targetHandle!).inputId === inputId
+        );
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [edgeChanges, getEdges, nodeId, inputId]);
+
+    const { functionDefinitions } = useContext(BackendContext);
+
+    const sourceTypeColor = useMemo(() => {
```

Comment:
Joey... did you actually just copy that entire color logic? Please extract all of this out into a hook or similar to avoid code duplications between this `InputHandle` and the real one.